cpp/src/tsne/barnes_hut_tsne.cuh

```diff
@@ -310,6 +310,14 @@ value_t Barnes_Hut(value_t* VAL,
                                                                      n);
     CUDA_CHECK(cudaPeekAtLastError());
 
+    value_t grad_norm =
+      compute_grad_norm(handle, old_forces.data(), old_forces.size());
+
+    if (grad_norm <= params.min_grad_norm) {
```

Comment:
I've wondered if this should be `if (iter > params.exaggeration_iter && grad_norm <= params.min_grad_norm)` so that it can't break during the exaggeration phase. CannyLab's doesn't do that, but it would seem to make sense. In my other comment, notice that the grad norm reaches almost 0 during the exaggeration phase before spiking at iteration 250.

(I never use early stopping. Every comment and plot I've ever posted to RAPIDSAI/cuML has been for a fixed number of iters.)